### PR TITLE
Fix CI, update py03 to latest version

### PR DESCRIPTION
**Pull request by notmandatory**
- Update ubuntu to 22.04 to fix GLIBC error, then fix OpenSSL ripemd160 config, see: https://stackoverflow.com/questions/74545780/how-to-enable-the-openssl-3-0-legacy-provider-github-actions
- Downgrade Leger emulator version (speculos), set version to 0.8.3
- Fix clippy warnings
- Update py03 dependency to 0.21.2

### CODE REVIEW

**Comment by ValuedMammal** [comment]
hm, I see speculos had a Traceback with `IndexError: bytearray index out of range` and I found what looks like a similar issue here https://github.com/LedgerHQ/speculos/issues/488#issuecomment-2124381715

**Comment by ValuedMammal** [comment]
I tried with a different firmware version in #102 with no luck. We could possibly try a different device than the nano s.

**Comment by notmandatory** [comment]
I got this to build with ubuntu 20.04 and rust 1.63 but with `stable` rust fails due to an error about requiring glibc 2.33. I think the `stable` rust version does work with ubuntu 22.04.  I suspect some dependency is updated with `stable` rust and we need to pin that. Interestingly `nightly` rust works fine with 20.04 (in the doc example test). I think when I tried with ubuntu 22.04 rust 1.63 and `nightly` failed.

**Comment by oleonardolima** [comment]
> I got this to build with ubuntu 20.04 and rust 1.63 but with `stable` rust fails due to an error about requiring glibc 2.33. I think the `stable` rust version does work with ubuntu 22.04. I suspect some dependency is updated with `stable` rust and we need to pin that. Interestingly `nightly` rust works fine with 20.04 (in the doc example test). I think when I tried with ubuntu 22.04 rust 1.63 and `nightly` failed.

@notmandatory I was researching this and the failure started to happen on version 0.8.4 further, the `IndexError: bytearray index out of range` error seems consistent on all versions from 0.8.4 up to 0.9.1.

In the PR https://github.com/bitcoindevkit/rust-hwi/pull/103 I was playing with the versions, it was only failing due to the glibc error on 1.78 (stable), but it works on previous ones, and in the nightly as you've mentioned (1.80).

It's weird, because I've tried to reproduce the failure on stable locally and wasn't able to, and now it worked on the CI 😅

**Comment by notmandatory** [comment]
Got it all working! updated description with details.

**Review by ValuedMammal — APPROVED** [review]
ACK a714626130b3d19502714676ed928d1613309c82

File: ci/Dockerfile.ledger [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM ghcr.io/ledgerhq/speculos
+FROM ghcr.io/ledgerhq/speculos:sha-f430e51
```
Comment (oleonardolima) [inline ci/Dockerfile.ledger:1]: @notmandatory Any reason for using this specific hash ?

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -59,17 +59,8 @@ jobs:
...
-        uses: docker/build-push-action@v3
-        with:
-          context: ci/
-          file: ci/Dockerfile.${{ matrix.emulator.name }}
-          tags: hwi/${{ matrix.emulator.name }}_emulator:latest
-          load: true
-          cache-from: type=gha
-          cache-to: type=gha
+      - name: Build simulator image
+        run: docker build -t hwi/${{ matrix.emulator.name }}_emulator:latest ./ci -f ci/Dockerfile.${{ matrix.emulator.name }}
```
Comment (oleonardolima) [inline .github/workflows/main.yml:63]: @notmandatory Was this required to make it work, or just an enhancement?

File: ci/Dockerfile.ledger (hunk 1 above)
Comment (oleonardolima) [inline ci/Dockerfile.ledger:1]: @notmandatory Any reason for using this specific hash? As per my tests, the most recent version that's working properly is 0.8.3.

File: .github/workflows/main.yml (hunk 2 above)
Comment (notmandatory) [reply .github/workflows/main.yml:63]: Not sure if this is required but this is how it was done in the job that runs the doc tests and it looked like a simpler way to do it so adopted here too.

File: ci/Dockerfile.ledger (hunk 1 above)
Comment (notmandatory) [reply ci/Dockerfile.ledger]: no reason I just picked one I thought was around the time we last had this working. If you have it working with 0.8.3 then let's use that one. I updated this PR to test it out too and 0.8.3 looks fine.

**Pull request merged**